app/src/main/java/org/mozilla/fenix/library/downloads/DownloadView.kt

```diff
@@ -55,18 +73,44 @@ class DownloadView(
     }
 
     fun update(state: DownloadFragmentState) {
+        val oldMode = mode
 
+        view.progress_bar.isVisible = state.isDeletingItems
         view.swipe_refresh.isEnabled = false
         mode = state.mode
 
-        updateEmptyState(state.items.isNotEmpty())
+        downloadAdapter.updatePendingDeletionIds(state.pendingDeletionIds)
+
+        updateEmptyState(state.pendingDeletionIds.size != state.items.size)
 
         downloadAdapter.updateMode(state.mode)
         downloadAdapter.updateDownloads(state.items)
 
-        setUiForNormalMode(
-            context.getString(R.string.library_downloads)
-        )
+        if (state.mode::class != oldMode::class) {
+            interactor.onModeSwitched()
+        }
+
+        if (state.mode is DownloadFragmentState.Mode.Editing) {
+            val unselectedItems = oldMode.selectedItems - state.mode.selectedItems
+
+            state.mode.selectedItems.union(unselectedItems).forEach { item ->
+                val index = state.items.indexOf(item)
+                downloadAdapter.notifyItemChanged(index)
+            }
+        }
+
+        when (val mode = state.mode) {
+            is DownloadFragmentState.Mode.Normal -> {
+                setUiForNormalMode(
+                    context.getString(R.string.library_downloads)
+                )
+            }
+            is DownloadFragmentState.Mode.Editing -> {
+                setUiForSelectingMode(
+                    context.getString(R.string.download_multi_select_title, mode.selectedItems.size)
+                )
+            }
+        }
     }
 
     fun updateEmptyState(userHasDownloads: Boolean) {
```

Comment:
This method name is confusing, because it is called when the state is not necessarily empty. I would rename this to be more clear on it's intention -- "handle visibility" or something